packages/ra-core/src/ui/input/ReferenceArrayInput.js

```diff
@@ -206,100 +177,49 @@ export class ReferenceArrayInput extends Component {
             input,
-            resource,
-            label,
-            source,
             referenceRecords,
-            allowEmpty,
             matchingReferences,
-            basePath,
             onChange,
             children,
-            meta,
-            record,
-            options,
             translate,
-            ...rest
         } = this.props;
 
-        if (React.Children.count(children) !== 1) {
-            throw new Error(
-                '<ReferenceArrayInput> only accepts a single child (like <Datagrid>)'
-            );
-        }
-
-        const translatedLabel = translate(
-            label || `resources.${resource}.fields.${source}`
-        );
-
         const dataStatus = getDataStatus({
             input,
             matchingReferences,
             referenceRecords,
             translate,
         });
 
-        if (dataStatus.waiting) {
-            return (
-                <Labeled
-                    label={translatedLabel}
-                    source={source}
-                    resource={resource}
-                    className={className}
-                    {...sanitizeRestProps(rest)}
-                >
-                    <LinearProgress />
-                </Labeled>
-            );
-        }
-
-        if (dataStatus.error) {
-            return (
-                <ReferenceError
-                    label={translatedLabel}
-                    error={dataStatus.error}
-                />
-            );
-        }
-
-        return React.cloneElement(children, {
-            className,
-            allowEmpty,
-            input,
-            label: translatedLabel,
-            resource,
-            meta: {
-                ...meta,
-                helperText: dataStatus.warning || false,
-            },
-            source,
+        return children({
             choices: dataStatus.choices,
-            basePath,
+            error: dataStatus.error,
+            isLoading: dataStatus.waiting,
             onChange,
             setFilter: this.debouncedSetFilter,
             setPagination: this.setPagination,
             setSort: this.setSort,
-            translateChoice: false,
-            options,
-            ...sanitizeRestProps(rest),
+            warning: dataStatus.warning,
         });
     }
 }
 
-ReferenceArrayInput.propTypes = {
+CoreReferenceArrayInput.propTypes = {
     allowEmpty: PropTypes.bool.isRequired,
     basePath: PropTypes.string,
-    children: PropTypes.element.isRequired,
+    children: PropTypes.func.isRequired,
     className: PropTypes.string,
     crudGetMatching: PropTypes.func.isRequired,
     crudGetMany: PropTypes.func.isRequired,
     filter: PropTypes.object,
     filterToQuery: PropTypes.func.isRequired,
     input: PropTypes.object.isRequired,
     label: PropTypes.string,
-    matchingReferences: PropTypes.array,
+    matchingReferences: PropTypes.oneOfType([
+        PropTypes.array,
+        PropTypes.object,
```

Comment:
in which case is it an object? Why did you change it in this PR?

Reply:
Because it was actually missing. A warning was displayed in the console when testing because it can become: `matchingReferences: { error: 'fetch error' }`

Checkout the unit tests